ambry-router/src/main/java/com/github/ambry/router/ParanoidDurabilityOperationTracker.java

```diff
@@ -0,0 +1,314 @@
+          return false;
+        }
+      } else {
+        if (replicaPoolByDc.get(currentDc).size() < remoteReplicaSuccessTarget) {
+          logger.error("Not enough replicas in remote data center " + currentDc + " for partition " + partitionId
+              + " to satisfy paranoid durability requirements " +
+              "(wanted " + remoteReplicaSuccessTarget + ", but found " + replicaPoolByDc.get(currentDc).size() + ")");
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public boolean hasSucceeded() {
+    return hasSucceededLocally() && hasSucceededRemotely();
+  }
+
+  private boolean hasSucceededLocally() {
+    return localReplicaSuccessCount >= localReplicaSuccessTarget;
+  }
+
+  private boolean hasSucceededRemotely() {
+    return remoteReplicaSuccessCount >= remoteReplicaSuccessTarget;
+  }
+
+  /**
+   * Add a replica to the beginning of the replica pool linked list for the given data center.
+   * @param replicaId the replica to add.
+   */
+  private void addToBeginningOfPool(ReplicaId replicaId) {
+    modifyReplicasInPoolOrInFlightCount(replicaId, 1);
+    replicaPoolByDc.computeIfAbsent(replicaId.getDataNodeId().getDatacenterName(), k -> new LinkedList<>())
+        .addFirst(replicaId);
+  }
+
+
+  /**
+   * Add a replica to the end of the replica pool linked list for the given data center.
+   * @param replicaId the replica to add.
+   */
+  private void addToEndOfPool(ReplicaId replicaId) {
+    modifyReplicasInPoolOrInFlightCount(replicaId, 1);
+    replicaPoolByDc.computeIfAbsent(replicaId.getDataNodeId().getDatacenterName(), k -> new LinkedList<>())
+        .addLast(replicaId);
+  }
+
+  public void onResponse(ReplicaId replicaId, TrackedRequestFinalState trackedRequestFinalState) {
+    super.onResponse(replicaId, trackedRequestFinalState);
+    modifyReplicasInPoolOrInFlightCount(replicaId, -1);
+
+    String dcName = replicaId.getDataNodeId().getDatacenterName();
+
+    switch (trackedRequestFinalState) {
+      case SUCCESS:
+        replicaSuccessCountByDc.put(dcName, replicaSuccessCountByDc.get(dcName) + 1);
+        if (dcName.equals(datacenterName)) {
+          localReplicaSuccessCount++;
+        } else {
+          remoteReplicaSuccessCount++;
+        }
+        break;
+      case REQUEST_DISABLED:
+        disabledCountByDc.put(dcName, disabledCountByDc.get(dcName) + 1);
+        break;
+      default:
+        failedCountByDc.put(dcName, failedCountByDc.get(dcName) + 1);
+    }
+  }
+
+  public boolean hasFailed() {
+    if (routerConfig.routerPutUseDynamicSuccessTarget) {
+      for(String dcName : allDatacenters) {
+        if (dcName.equals(datacenterName)) {  // Local colo
+          if (hasFailedDynamically(dcName, routerConfig.routerPutSuccessTarget)) return true;
+        } else {
+          if (hasFailedDynamically(dcName, routerConfig.routerPutRemoteSuccessTarget)) return true;
```

Comment:
the dynamic target should only apply to local dc.

Reply:
👍 agreed and changed.